### PR TITLE
User can store it's company data

### PR DESCRIPTION
**Pull request by EevanR**
[PT](https://www.pivotaltracker.com/story/show/170302032)

User can store his/her own company's data
- Companies controller and company model with appropriate view page
- Input fields for storing company data
- show page showing saved company info
- redirect from company index based on whether saved company info exists or not 
- corresponding feature tests

### CODE REVIEW

File: features/user_can_store_its_company_data.feature [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    Given the following user exist
+      | email         | password |
+      | user@user.com | password |
+
+    And I am logged in as "user@user.com"  
+    And I am on the index page
+    
+  Scenario: Visitor can save his company information
+    When I click on "Company Profile"
+    Then I should be on the Company Profile page
```
Comment (oliverochman) [inline features/user_can_store_its_company_data.feature]: a redundant assertion, please remove

File: features/user_can_store_its_company_data.feature [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    And I should see "Name"
+    And I fill in "Name" with "Bob"
+    And I click on "Update profile"
+    Then I should be on the Invoices page
+    And I should see "Company information added."
+
+
+  Scenario: Company Name must be filled in
+    When I click on "Company Profile"
+    Then I should be on the Company Profile page
```
Comment (oliverochman) [inline features/user_can_store_its_company_data.feature]: a redundant assertion, please remove

File: features/user_can_store_its_company_data.feature [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    And I click on "Update profile"
+    Then I should see "Something went wrong. Company Name is required."
+
+  Scenario: User can view saved company information
+    When I click on "Company Profile"
+    Then I should be on the Company Profile page
```
Comment (oliverochman) [inline features/user_can_store_its_company_data.feature:38]: a redundant assertion, please remove

File: features/user_can_store_its_company_data.feature [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    And I fill in "Name" with "Bob"
```
Comment (oliverochman) [inline features/user_can_store_its_company_data.feature]: please add a proper company name

File: spec/models/company_spec.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    it { is_expected.to have_db_column :vat_no }
+    it { is_expected.to have_db_column :address }
+    it { is_expected.to have_db_column :postcode }
+    it { is_expected.to have_db_column :city }
+    it { is_expected.to have_db_column :phone }    
+    it { is_expected.to have_db_column :user_id }
+  end
+
+  describe 'Associations' do
+    it { is_expected.to belong_to :user }
```
Comment (oliverochman) [inline spec/models/company_spec.rb:16]: you need to test the association in the user unit spec as well

File: app/controllers/companies_controller.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
+class CompaniesController < ApplicationController
+  before_action :authenticate_user!
+  
+  def index
```
Comment (oliverochman) [inline app/controllers/companies_controller.rb]: why do you have an index action for companies if a user only `has_one`

File: app/controllers/companies_controller.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    company = current_user.company
+    if company == nil
```
Comment (oliverochman) [inline app/controllers/companies_controller.rb]: this conditional should be in the show action

File: app/controllers/companies_controller.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      redirect_to new_company_path
+    else
+      redirect_to company_path(Company.ids)
```
Comment (oliverochman) [inline app/controllers/companies_controller.rb]: can you explain this line to me

File: app/models/company.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+class Company < ApplicationRecord
+  belongs_to :user
+  validates :name, :presence => true
```
Comment (oliverochman) [inline app/models/company.rb]: please use `validates_presence_of:`

File: app/views/layouts/_nav.html.erb [hunk 10]
```diff
@@ -1,6 +1,7 @@
 <nav>
   <% if user_signed_in? %>
     <%= link_to 'EXTREME INVOICING', invoices_path %>
+    <%= link_to 'Company Profile', companies_path %>
```
Comment (oliverochman) [inline app/views/layouts/_nav.html.erb]: this should lead directly to the company show path

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: features/user_can_store_its_company_data.feature (hunk 1 above)
Comment (EevanR) [reply features/user_can_store_its_company_data.feature]: done

File: features/user_can_store_its_company_data.feature (hunk 2 above)
Comment (EevanR) [reply features/user_can_store_its_company_data.feature]: done

File: features/user_can_store_its_company_data.feature (hunk 3 above)
Comment (EevanR) [reply features/user_can_store_its_company_data.feature:38]: done

File: features/user_can_store_its_company_data.feature (hunk 4 above)
Comment (EevanR) [reply features/user_can_store_its_company_data.feature]: changed

File: app/models/company.rb (hunk 9 above)
Comment (EevanR) [reply app/models/company.rb]: changed

File: app/views/company/index.html.erb [hunk 11]
```diff
@@ -0,0 +1 @@
+<h2>Company Profile</h2>
```
Comment (oliverochman) [inline app/views/company/index.html.erb]: please remove this file

File: config/routes.rb [hunk 12]
```diff
@@ -2,5 +2,6 @@
   devise_for :users
   resources :invoices, only: [:index, :new]
   root controller: :landing, action: :index
+  resources :companies, only: [:index, :create, :show, :new]
```
Comment (oliverochman) [inline config/routes.rb]: do you need the index action?

File: features/user_can_store_its_company_data.feature [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    And I fill in "Name" with " "
+    And I click on "Update profile"
+    Then I should see "Something went wrong. Name can't be blank"
+
+
+  Scenario: User can view saved company information
+    Given I am logged in as "google_exec@google.com"  
+    And I am on the index page
+    When I click on "Company Profile"
+    And I should see "Google"
```
Comment (oliverochman) [inline features/user_can_store_its_company_data.feature]: Then I should see

File: app/controllers/companies_controller.rb [hunk 14]
```diff
@@ -0,0 +1,34 @@
+class CompaniesController < ApplicationController
+  before_action :authenticate_user!
+  before_action :check_user_company, only: :new
+
```
Comment (oliverochman) [inline app/controllers/companies_controller.rb:4]: remove the unnecessary empty line

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/controllers/companies_controller.rb (hunk 14 above)
Comment (EevanR) [reply app/controllers/companies_controller.rb:4]: done

File: features/user_can_store_its_company_data.feature (hunk 13 above)
Comment (EevanR) [reply features/user_can_store_its_company_data.feature]: done

File: config/routes.rb (hunk 12 above)
Comment (EevanR) [reply config/routes.rb]: no

File: app/views/company/index.html.erb (hunk 11 above)
Comment (EevanR) [reply app/views/company/index.html.erb]: done

**Review by oliverochman — APPROVED** [review]

**Pull request merged**